File: PHP/Patterns/src/Behavioral/Template/ExerciseOne/VideoEditingPreset.php
<?php

namespace App\Behavioral\Template;

abstract class VideoEditingPreset
{
    public function applyPreset(): void
    {
        $this->applyColorCorrection();
        $this->applyFilters();
        $this->enhanceVideoQuality();
        $this->enhanceAudioQuality();
        $this->renderVideo();
    }

    private function applyFilters(): void
    {
    }
    abstract protected function applyColorCorrection(): void;

    private function enhanceVideoQuality():void{}
    private function enhanceAudioQuality():void{}

    protected abstract function renderVideo():void;
}

File: C#/Template/Exercise/Abstruction/VideoEditingPreset.cs
namespace Template.Exercise.Abstruction;
internal abstract class VideoEditingPreset
{
    internal void ProcessVideo()
    {
        EnhanceVideoQuality();
        ApplyColorCorrection();
        EnhanceAudioQuality();
        ApplyFilters();

        RenderVideo();
    }

    private void EnhanceVideoQuality()
    {
        Console.WriteLine("Enhancing video quality...");
    }

    private void ApplyColorCorrection()
    {
        Console.WriteLine("Applying color correction...");
    }

    private void EnhanceAudioQuality()
    {
        Console.WriteLine("Enhancing audio quality...");
    }

    private void ApplyFilters()
    {
        Console.WriteLine("Applying filters...");
    }

    protected abstract void RenderVideo();
}
